core/include/drivers/clk.h

```diff
@@ -0,0 +1,172 @@
+	TEE_Result (*set_parent)(struct clk *clk, size_t index);
+	size_t (*get_parent)(struct clk *clk);
+	TEE_Result (*set_rate)(struct clk *clk, unsigned long rate,
+			       unsigned long parent_rate);
+	unsigned long (*get_rate)(struct clk *clk,
+				  unsigned long parent_rate);
+};
+
+/**
+ * Return the clock name
+ *
+ * @clk: Clock for which the name is needed
+ * Return a const char* pointing to the clock name
+ */
+static inline const char *clk_get_name(struct clk *clk)
+{
+	return clk->name;
+}
+
+/**
+ * clk_alloc - Allocate a clock structure
+ *
+ * @name: Clock name
+ * @ops: Clock operations
+ * @parent_clks: Parents of the clock
+ * @parent_count: Number of parents of the clock
+ *
+ * Returns a clock struct properly initialized or NULL if allocation failed
+ */
+struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
+		      struct clk **parent_clks, size_t parent_count);
+
+/**
+ * clk_free - Free a clock structure
+ *
+ * @clk: Clock to be freed or NULL
+ */
+void clk_free(struct clk *clk);
+
+/**
+ * clk_register - Register a clock within the clock framework
+ *
+ * @clk: Clock struct to be registered
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result clk_register(struct clk *clk);
+
+/**
+ * clk_get_rate - Get clock rate
+ *
+ * @clk: Clock for which the rate is needed
+ * Returns the clock rate in Hz
+ */
+unsigned long clk_get_rate(struct clk *clk);
+
+/**
+ * clk_set_rate - Set a clock rate
+ *
+ * @clk: Clock to be set with the rate
+ * @rate: Rate to set in Hz
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate);
+
+/**
+ * clk_enable - Enable a clock and its ascendance
+ *
+ * @clk: Clock to be enabled
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result clk_enable(struct clk *clk);
+
+/**
+ * clk_disable - Disable a clock
+ *
+ * @clk: Clock to be disabled
+ * Returns 0 on success or a negative value on error.
+ */
+void clk_disable(struct clk *clk);
+
+/**
+ * clk_get_parent - Get the current clock parent
+ *
+ * @clk: Clock for which the parent is needed
+ * Return the clock parent or NULL if there is no parent
+ */
+struct clk *clk_get_parent(struct clk *clk);
+
+/**
+ * clk_get_num_parents - Get the number of parent for a clock
+ *
+ * @clk: Clock for which the number of parents is needed
+ * Return the number of parents
+ */
+static inline size_t clk_get_num_parents(struct clk *clk)
+{
+	return clk->num_parents;
+}
+
+/**
+ * Get a clock parent by its index
+ *
+ * @clk: Clock for which the parent is needed
+ * @pidx: parent index for the clock
+ * Return the clock parent at index pidx or NULL if out of bound
+ */
+struct clk *clk_get_parent_by_index(struct clk *clk, size_t pidx);
+
+/**
+ * clk_set_parent - Set the current clock parent
+ *
+ * @clk: Clock for which the parent should be set
+ * @pidx: parent index to set
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result clk_set_parent(struct clk *clk, size_t pidx);
```

Comment:
In reply to comment [886819931](https://github.com/OP-TEE/optee_os/pull/4705#issuecomment-886819931):
> For good accuracy on requirements I think we need input from @etienne-lms too. (...)

All the generic things a clk driver layer can provide may be duplication of how the platform clock driver describes the hardware. I was thinking platform backend driver that do it its own way.

Here, it could be a config switch `CFG_DRIVERS_PLAT_CLK=y|n` using _clk.h_ as the shim to a `plat_clk_*()` platform driver interface.
```patch
@@@
 #ifndef __DRIVERS_CLK_H
 #define __DRIVERS_CLK_H
 
+#ifndef CFG_DRIVERS_PLAT_CLK
 #include <kernel/refcount.h>
 #include <stdint.h>
 #include <tee_api_types.h>

@@@
  * Returns a TEE_Result compliant value
  */
 TEE_Result clk_set_parent(struct clk *clk, size_t pidx);
+#else /* !CFG_DRIVERS_PLAT_CLK */
+/*
+ * CFG_DRIVERS_PLAT_CLK relies on a single memory optimized
+ * platform clock provider. Some features from clk.h are not supported,
+ * mainly clock parenthood description and configuration.
+ */
+
+/* Opaque reference struct clk, meaninful to plat_clk_*() functions */
+struct clk;
+
+/* Platform shall implement to following functions */
+char *plat_clk_get_name(struct clk *clk);
+struct clk *plat_clk_find_by_name(const char *name);
+bool plat_clk_is_valid(struct clk *clk);
+unsigned long plat_clk_get_rate(struct clk *clk);
+TEE_Result plat_clk_change_rate(struct clk *clk, unsigned long rate);
+TEE_Result plat_clk_enable(struct clk *clk);
+void plat_clk_disable(struct clk *clk);
+
+/* Clock driver helper functions route to plat_clk_*() when supported */
+static inline const char *clk_get_name(struct clk *clk)
+{
+	return plat_clk_get_name(clk);
+}
+
+static inline struct clk *clk_alloc(const char *name,
+				    const struct clk_ops *ops __maybe_unused,
+				    struct clk **parent_clks __maybe_unused,
+				    size_t parent_count __maybe_unused)
+{
+	assert(!ops && !parent_clks && !parent_count);
+
+	return plat_clk_find_by_name(name);
+}
+
+static inline void clk_free(struct clk *clk __unused)
+{
+}
+
+TEE_Result clk_register(struct clk *clk)
+{
+	if (plat_clk_is_valid(clk))
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static inline unsigned long clk_get_rate(struct clk *clk)
+{
+	return plat_clk_get_rate(clk);
+}
+
+static inline TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	return plat_clk_change_rate(clk, rate);
+}
+
+static inline TEE_Result clk_enable(struct clk *clk);
+{
+	return plat_clk_enable(clk);
+}
+
+static inline struct clk *clk_get_parent_by_index(struct clk *clk __unused,
+						  size_t pidx __unused);
+{
+	// maybe add plat_clk_parent_*() functions in plat_clk.h
+	return NULL;
+}
+
+static inline TEE_Result clk_set_parent(struct clk *clk__unused,
+					size_t pidx __unused);
+{
+	DMSG("Unpexpected call with CFG_DRIVERS_PLAT_CLK");
+	return TEE_ERROR_GENERIC;
+}
+
+static inline void clk_disable(struct clk *clk);
+{
+	plat_clk_disable(clk);
+}
+
+static inline struct clk *clk_get_parent(struct clk *clk __unused)
+{
+	return NULL;
+}
+
+static inline size_t clk_get_num_parents(struct clk *clk __unused)
+{
+	return 0;
+}
+#endif /* !CFG_DRIVERS_PLAT_CLK */
-
 #endif /* __DRIVERS_CLK_H */
```

But I think this approach is why @jenswi-linaro mentioned as the failure option 3 in https://github.com/OP-TEE/optee_os/pull/4705#issuecomment-886414835.